Add a way to get the source of a primed TNT entity

MisterVector opened the pull request:
Since Minecraft 1.5, we can now get the source of a primed TNT entity, so
we just reflect those changes in bukkit

CB PR: https://github.com/Bukkit/CraftBukkit/pull/1065

JIRA: https://bukkit.atlassian.net/browse/BUKKIT-3815

authorblues commented:
Squash.

MisterVector commented:
It has been squashed.

Wolvereness commented:
What if fire arrow ignites TNT? That's not living...

Wolvereness commented on src/main/java/org/bukkit/entity/TNTPrimed.java:
```diff
@@ -17,4 +17,15 @@
+     * responsible for the creation of this primed TNT.
+     * (I.E. player ignites TNT with flint and steel)
+     * Take note that this can be null if there is no suitable source.
+     * (created by the {@link org.bukkit.World#spawn(Location, Class)} method)
```
78 characters for javadoc, plus newline.

MisterVector commented:
Well, the source field in EntityTNTPrimed is actually an EntityLiving, so there's no way an arrow could be a source with that...

Wolvereness commented:
Not saying it's possible / impossible now, but for the future. I don't agree with making it strictly Living.

MisterVector commented:
In my opinion it shouldn't really be living either. For example, the base Entity class has a setPassenger() method, instead of it being given to a subclass like EntityLiving.

MisterVector commented:
There, spaced it out, so it should be good now. :)

MisterVector commented:
Now I use the base Entity class, to allow for extensibility if Mojang decides to make it an Entity instead of EntityLiving.

jrtc27 commented on src/main/java/org/bukkit/entity/TNTPrimed.java:
```diff
@@ -17,4 +17,15 @@
      * @return the number of ticks until this TNTPrimed explodes
      */
     public int getFuseTicks();
+    /**
```
Should be a blank line above this

MisterVector commented:
Thank you. Fixed.

MisterVector commented:
Added some notes that the source becomes null when the chunk it's in is unloaded.

Wolvereness commented:
Does that include if a player gets teleported or logs off / back on but the chunk doesn't get unloaded?

MisterVector commented:
If the chunk doesn't get unloaded, the source will still contain that player reference. When I had a player nearby, I tested this out by placing TNT, igniting it, logging off, then quickly logging back in before it could explode.

Wolvereness commented:
I really think this needs some way of _not_ doing that... Stale references are evil.

md-5 commented:
Dun dun dun WeakReference

MisterVector commented:
Added some more information. I am satisfied with this PR at this point.

MisterVector committed:
Add a way to get the source of a primed TNT entity

Since Minecraft 1.5, we can now get the source of a primed TNT entity, so
we just reflect those changes in bukkit

MisterVector committed:
- Added some more javadoc for an additional reason the source may be null

natemort commented:
Pulled.
